### PR TITLE
Resolves #84 - new rule to prevent circular project dependencies

### LINKED ISSUE #84: New rule to prevent circular dependencies in a maven project (not just a module)
It is possible to have each module have a set of depencies which are not circular in themselves but which makes it impossible to build either project, i.e. 
Suppose you have two 'projects' (A, B) each of which has two modules (1, 2). 
A2 depends on A1, B1
B2 depends on A1, B1
Therefore I cannot build A1 and A2 without B1, but I cannot build B1, B2 without A1.
I propose a new rule `BanProjectCircularDependencies` which prevents this.

### PR DESCRIPTION
**Pull request by harrisric**
Code and test for a new rule.

### CODE REVIEW

File: src/it/circular-project/invoker.properties [hunk 1]
```diff
@@ -0,0 +1,3 @@
+invoker.goals=enforcer:enforce
+invoker.buildResult = failure
```
Comment (pzygielo) [inline src/it/circular-project/invoker.properties]: This IT fails indeed, but I'm not sure it happens in expected way:
```
[ERROR] Child module .../target/it/circular-project/slf4j-api of .../target/it/circular-project/pom.xml does not exist @ 
[ERROR] 'packaging' with value 'jar' is invalid. Aggregator projects require 'pom' as packaging. @ line 2, column 106
 @ 
[ERROR] The build could not read 1 project -> [Help 1]
```

File: src/it/circular-project/pom.xml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>org.slf4j</groupId>
+  <artifactId>slf4j-parent</artifactId>
+  <version>1.6.1</version>
+  <name>CircularProjectDependencies test</name>
+
+  <modules>
+    <module>slf4j-api</module>
```
Comment (pzygielo) [inline src/it/circular-project/pom.xml]:
```suggestion
    <module>child-a</module>
```

File: src/it/circular-project/pom.xml [hunk 3]
```diff
@@ -0,0 +1,36 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>org.slf4j</groupId>
+  <artifactId>slf4j-parent</artifactId>
+  <version>1.6.1</version>
+  <name>CircularProjectDependencies test</name>
+
```
Comment (pzygielo) [inline src/it/circular-project/pom.xml]:
```suggestion
  <packaging>pom</packaging>

```

File: src/it/circular-project/invoker.properties (hunk 1 above)
Comment (pzygielo) [reply src/it/circular-project/invoker.properties]: With changes as suggested for `it/.../pom.xml` - build of `circular-project` completes successfully, (which is test failure, as `banProjectCircularDependencies` is not triggered).

Comment (harrisric) [reply src/it/circular-project/invoker.properties]: Thanks - it seems I've been misled by the local build - you are correct that it seems to be failing for a different reason (the pom-packaging) and when that is corrected there is an error being logged which indicates that the trick of trying to re-use an existing project isn't working:
[WARNING] The POM for org.slf4j:jcl-over-slf4j:jar:1.6.1 is invalid, transitive dependencies (if any) will not be available: 2 problems were encountered while building the effective model for org.slf4j:jcl-over-slf4j:1.6.1
[ERROR] 'dependencies.dependency.version' for org.slf4j:slf4j-api:jar is missing. @ 
[ERROR] 'dependencies.dependency.version' for org.slf4j:slf4j-jdk14:jar is missing. @ 

Seems that its a bit tricky to set up a good test case for this. Will have another look and see if I can set up projects directly for the test.

Comment (harrisric) [reply src/it/circular-project/invoker.properties]: I've added specific version dependencies to the pom for the it and the build now fails correctly because of the rule with message:
[WARNING] Rule 0: org.apache.maven.plugins.enforcer.BanProjectCircularDependencies failed with message:
Circular Project Dependency found. No modules groupId:artifactId combination in your project must exist in the list of transitive dependencies for any module in the project.
  org.slf4j:slf4j-api found in transitive dependencies of org.slf4j:childb

Comment (pzygielo) [reply src/it/circular-project/invoker.properties]: > the build now fails correctly because of the rule with message:

Perhaps this could be verified during the build? For example https://github.com/mojohaus/extra-enforcer-rules/blob/52017dd8ff5e9ed7356c4e8f4d52ba070a1ba03f/src/it/mojo-1853/invoker.properties#L4 also is expecting build failure and in https://github.com/mojohaus/extra-enforcer-rules/blob/52017dd8ff5e9ed7356c4e8f4d52ba070a1ba03f/src/it/mojo-1853/verify.groovy#L4 the reason seems to be checked.

Just to make sure that only the specific failure is accepted as success, and not _any_ failure.

Comment (harrisric) [reply src/it/circular-project/invoker.properties]: Great idea, thanks - will add that.

**Comment by harrisric** [comment]
@pzygielo is there something that still needs to be done or can this change be merged?

**Comment by pzygielo** [comment]
Hi @harrisric - I have no power over this repository.
Perhaps @rfscholte could suggest how to proceed.

**Comment by rfscholte** [comment]
Maybe it is me, but I don't see a circular dependency here. And it doesn't have to be an issue. Circular dependencies at dependencies are impossible, but at runtime they are valid. This requires more clarification.

**Comment by harrisric** [comment]
@rfscholte the scenario that I have is where we have two maven projects using the same versioning sequence.
So project A has modules A-1, A-2, project B has modules B-1. A-1 has a dependency on B-1
Both have a built version 1.0.0 and have a current build version of 1.0.1-SNAPSHOT 
We now introduce a dependency from B-1 to A-2. We continue to be able to build projects A and B with version 1.0.1-SNAPSHOT  because that version already exists for the other project.
If I now remove the snapshot to formalise the release of 1.0.1 neither project A or B can build because that version of the other project (which is needed for one of its dependencies) does not yet exist.

**Comment by harrisric** [comment]
@rfscholte is there anything further I can do to help clarify here?

**Comment by gmshake** [comment]
@harrisric 
Looks like your scenario is caused by releasing progress circular dependency.
If snapshot dependency is not permitted in release version, then at lease three releasing are required.

Suppose initial version of project A is 1.0.0 and project B is 2.0.0, try the following releasing progress.

1. Bump version of project A, 1.0.0 -> 1.0.1-SNAPSHOT
2. Make changes to A-2 that required by B-1
3. Release project A.  ( 1.0.1-SNAPSHOT -> 1.0.1 )
4. Bump the version of project B ( 2.0.0 -> 2.0.1-SNAPSHOT )
5. Introduce a dependency from B-1 to A-2(1.0.1)
6. Release project B. ( now is 2.0.1 )
7. Bump version of project A. ( 1.0.1 -> 1.0.2-SNAPSHOT )
8. Update the version of dependency of B-1 from 2.0.0 to  2.0.1 
9. Release project A. ( now is 1.0.2 )

Simply put, since the dependency is A1 -> B1 -> A2, release reversely.

**Comment by rfscholte** [comment]
@gmshake  is right: what you describe is not a circular dependency.
A circular dependency would be A1 -> B -> A1
However you say A1 -> B -> A2
When building B, Maven cannot see that A1 and A2 are part of the same multimodule project.
I don't think a new rule will help here.

**Comment by harrisric** [comment]
@rfscholte , @gmshake  I completely agree - where you are managing the versions of the two projects independently then you can manage this, though i think that the 9 step plan shows how painful it is for a single change (and that it specifically requires you to increment versions when this sort of change is identified).
In practice I am involved in a codebase where multiple maven projects are versioned together and new versions are not generated from individual changes but periodically to include new sets of features and fixes. As each maven project is built and published separately, the sort of 'cycle' between projects causes a block which requires manual identification and intervention.
This sort of rule works for us internally but I completely understand if you don't think that this is generally applicable. If so, please feel free to close this request.

**Comment by rfscholte** [comment]
Based on last comment of @harrisric and my opinion we should close this as won't fix
